### ISSUE TITLE
Expensive construction of `basic_string` from `forward_range`

### ISSUE THREAD

**Issue by AlexBAV** [microsoft/STL#3197]
When `basic_string` range constructor is invoked with a `forward_range` (for example via `ranges::to`), the range is traversed twice: first in order to get a number of elements and then in order to construct elements:

https://github.com/microsoft/STL/blob/ea092540b4a3347947863a447d820a9905a68c5b/stl/inc/xstring#L2872

I wonder if it is a deliberate decision? 

While the current implementation traverses the range twice and makes one allocation, the alternative would be to traverse once, but make more allocations. The latter may still be better for small strings (when no allocations are actually needed), for very large strings (when asymptotically allocations will not matter much) and especially for expensive lazy range processing.

**Comment by AlexBAV** [comment]
The same question is for `vector` range constructor, which also traverses forward ranges twice:

https://github.com/microsoft/STL/blob/ea092540b4a3347947863a447d820a9905a68c5b/stl/inc/vector#L728

**Comment by AlexBAV** [comment]
I understand. The problem is that benchmarking would need to be done in each particular case, because STL cannot tell how expensive the range processing is. The problem is that Standard does not provide us with a way to pass additional information to help STL to choose a correct strategy.

I see two solutions for this problem (of course it is relevant for cases of very expensive range processing, like expensive `ranges::filter`, for example):

1. Introduce a new range adaptor `ranges::as_input` (similar `ranges::as_rvalue` or `ranges::as_const`) to "convert" forward range into input range.
2. Add new constructor to `vector`, taking an estimated total number of elements (like `vector::vector(from_range_t, Range &&range, size_t estimated_total_elements)`). In this case we would be able to do the following:

```C++
auto vector = source_data | std::views::filter(/* expensive filter here */) | std::ranges::to<std::vector>(estimated_number_of_elements);
```

Vector constructor will do a `reserve` and traverse the range only once, possibly trading extra allocated memory for speed.

Bute surely those are out of scope here.
